0011-resharding-without-downtime.md

```diff
@@ -0,0 +1,195 @@
+    // digest, weight, key and hash_initialization of the configuration.
+    //
+    // When removing a shard from the map it is guaranteed that only blobs
+    // which resolved to the removed shard will get a different shard. When
+    // adding shards there is a weight/total_weight probability that any given
+    // blob will be resolved to the new shards.
+    map<string, Shard> shard_map = 3;
+}
+```
+
+Other algorithms considered were: [Consistent
+hashing](https://en.wikipedia.org/wiki/Consistent_hashing) and
+[Maglev](https://storage.googleapis.com/gweb-research2023-media/pubtools/2904.pdf).
+
+## Handling Topology Changes Dynamically
+
+To reduce the amount of restarts required a new blob access configuration is
+added.
+
+```
+message RemotelyDefinedBlobAccessConfiguration {
+    // Fetch the blob access configuration from an external service
+    buildbarn.configuration.grpc.ClientConfiguration grpc = 1;
+
+    // Maximum grace time when receiving a new blob access configuration
+    // before cancelling ongoing requests.
+    //
+    // Recommended value: 0s for client facing systems, 60s for internal
+    // systems
+    google.protobuf.Duration maximum_grace_time = 2;
+}
+```
+
+Which calls to a service that implements the following:
+
+```
+service RemoteBlobAccessConfiguration {
+    rpc Synchronize(SynchronizeRequest) returns (SynchronizeResponse);
+}  
+
+message SynchronizeRequest {
+    enum StorageBackend {
+        CAS = 0;
+        AC = 1;
+        ICAS = 2;
+        ISCC = 3;
+        FSAC = 4;
+    }
+    // An implementation defined identifier describing who the client is which
+    // the remote service may take into consideration when returning the
+    // BlobAccessConfiguration. This is typically used when clients are in
+    // different networks and should route differently.
+    string identifier = 1;
+
+    // Which storage backend that the service should describe the topology for.
+    StorageBackend storage_backend = 2;
+
+    // A message describing the current state from the perspective of the
+    // client, the client may assume it's current topology is correct if the
+    // service does not respond with a request to change it.
+    BlobAccessConfiguration current = 3;
+}
+
+message SynchronizeResponse {
+    // A message describing how the component should perform blob access
+    // configurations.
+    BlobAccessConfiguration desired_state = 1;
+    
+    // Latest time at which an ongoing request to the previous blob access
+    // configuration should be allowed to continue before the caller should
+    // cancel it and return an error. The client may cancel it before the
+    // grace_time has passed.
+    google.protobuf.Timestamp grace_time = 2;
+
+    // Timestamp for when this state is considered expired. The remote blob
+    // access configuration will avoid breaking compatibility until
+    // after this timestamp has passed.
+    google.protobuf.Timestamp response_expiry = 3;
```

Comment:
I don't understand the description. What happened when `response_expiry` has passed? Should an empty configuration be used then?

Reply:
It is to give some leeway in case the client can not reach the RemoteConfigurator where it can still consider itself valid. The component should consider itself partitioned from the cluster in case it can not contact the RemoteConfigurator within that time window (and I guess shutdown and/or return unavailable?).